Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,789 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+import Atomics
+import NIOConcurrencyHelpers
+
+/// The delegate of the ``NIOAsyncWriter``.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public protocol NIOAsyncWriterDelegate: Sendable {
+    /// The `Element` type of the delegate and the writer.
+    associatedtype Element: Sendable
+    /// The `Failure` type of the delegate and the writer.
+    associatedtype Failure: Error = Never
+
+    /// This method is called once a sequence was yielded to the ``NIOAsyncWriter``.
+    ///
+    /// If the ``NIOAsyncWriter`` was writable when the sequence was yielded, the sequence will be forwarded
+    /// right away to the delegate. If the ``NIOAsyncWriter`` was _NOT_ writable then the sequence will be buffered
+    /// until the ``NIOAsyncWriter`` becomes writable again.
+    func didYield<S: Sequence>(contentsOf sequence: S) where S.Element == Element
+
+    /// This method is called once the ``NIOAsyncWriter`` is terminated.
+    ///
+    /// Termination happens if:
+    /// - The ``NIOAsyncWriter`` is deinited.
+    /// - ``NIOAsyncWriter/finish()`` is called.
+    /// - ``NIOAsyncWriter/finish(with:)`` is called.
+    ///
+    /// - Note: This is guaranteed to be called _exactly_ once.
+    ///
+    /// - Parameter failure: The failure that terminated the ``NIOAsyncWriter``. If the writer was terminated without an
+    /// error this value is `nil`.
+    func didTerminate(failure: Failure?)
+}
+
+/// Errors thrown by the ``NIOAsyncWriter``.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncWriterError: Error, Hashable {
+    @usableFromInline
+    internal enum _Code: Hashable, Sendable {
+        case alreadyFinished
+    }
+
+    @usableFromInline
+    let _code: _Code
+
+    @usableFromInline
+    var file: String
+
+    @usableFromInline
+    var line: Int
+
+    @inlinable
+    init(_code: _Code, file: String = #fileID, line: Int = #line) {
```

Comment:
We should remove the defaults here so we ensure they're always provided by the helpers